Add alerting & recording rules, series deletion & retention docs (#397)
timescale/promscale: docs/alerting-recording.md

# Alerting & Recording Rules

### Alerting Rules

> The alerting rules are used to trigger alerts based on the provided alerting rules an alert should be fired to an external service like slack. In promscale the alerting rules are evaluated in prometheus. i.e define alerting rules in prometheus config

Reviewer: 1. This sentence does not sound good.
2. We do not define alerting rules in the prometheus configuration file. We mention the alerting-rules file name as a list where these files contains the actual rules. This statement needs an update in this direction.

Reviewer: ```suggestion
Alerting rules are used to trigger alerts based on the violation of any condition(s). These alerts are fired to external services like slack, mails, etc by the alert manager (link). Alerting rules are written in a YAML file and paths of these files are mentioned in the Prometheus configuration respectively. It is important to note that the evaluation of these conditional rules are done on the Prometheus side.
```

```
groups:
  - name: <alert-group-name>
    rules:
    - alert: <alert-name>
      expr: <promql_expression>
      for: <time-interval for how long this to happen to happen to fire an alert>
      labels:
        <key>:<value>
      annotations:
        summary: <text>
        description: <description on alert>
```

> Load the above defined alert rule file to your prometheus config for evaluation. i.e 

Reviewer: Maybe, have this statement linked with the above one after updating.

Author: The sequence is to add alerting rules in a file, define a reference to the rules file, and configure the alert manager endpoint.

Reviewer: Can we please change `config` -> `configuration` and `i.e` -> `i.e.,` (along with full-stops where needed) across the entire file?

```
rule_files:
  - "<alert-rules-file>"
```

Configuring the alert manager in prometheus config

```
alerting:
  alertmanagers:
  - static_configs:
    - targets:
      - localhost:9093
```

### Recording Rules

> The recording rules are for pre-calculating frequently used or computationally expensive queries. The results of those rules are saved into their own time series. In promscale the rules rules are evaluated in prometheus. i.e define recording rules in prometheus config

Reviewer: please mention that these newly formed time-series follow the name of the record (please confirm once)

# Alerting & Recording Rules

### Alerting Rules

> Alerting rules are used to trigger alerts based on the violation of any condition(s). These alerts are fired to external services like slack, mails, etc by the alert manager (link). Alerting rules are written in a YAML file and paths of these files are mentioned in the Prometheus configuration respectively. It is important to note that the evaluation of these conditional rules are done on the Prometheus side.

Reviewer: ```suggestion
Alerting rules are used to trigger alerts based on the violation of any condition(s). These alerts are fired to external services like slack, mails, etc by the alert manager (link). Alerting rules are written in a YAML file and paths of these files are mentioned in the Prometheus configuration respectively. It is important to note that the evaluation of these conditional rules are done at the Prometheus side.
```

1. Does this sound better with `at`?
2. Please add the link to the alert manager.

> Alerting rules are used to trigger alerts based on the violation of any condition(s). These alerts are fired to external services like slack, mails, etc by the [alert manager](https://prometheus.io/docs/alerting/latest/alertmanager/). Alerting rules are written in a YAML file and paths of these files are mentioned in the Prometheus configuration respectively. It is important to note that the evaluation of these conditional rules are done at the Prometheus side.

Reviewer 2: We should clarify whether this will fetch data from promscale if needed or not

```
groups:
  - name: <alert-group-name>
    rules:
    - alert: <alert-name>
      expr: <promql_expression>
      for: <time-interval for how long this to happen to happen to fire an alert>
      labels:
        <key>:<value>
      annotations:
        summary: <text>
        description: <description on alert>
```

Load the above defined alert rule file to your prometheus configuration for evaluation.

```
rule_files:
  - "<alert-rules-file>"
```

Configure the alert manager in prometheus configuration.

```
alerting:
  alertmanagers:
  - static_configs:
    - targets:
      - localhost:9093
```

More details on alerting rules can be found [here](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/).

### Recording Rules

> The recording rules are for pre-calculating frequently used or computationally expensive queries. The results of those rules are saved into their own time series. The newly formed time-series will follow the name of the record. In promscale the rules are evaluated in prometheus i.e define recording rules in prometheus configuration.

Reviewer 2: these rules are fired in prometheus but will they fetch data from promscale if needed or does the data retention in prometheus have to be long enough to contain all the data. we should clarify.

Also we should clarify that the new series is written into promscale.

Author: If `read_recent: true` is configured under `remote-read` configuration of prometheus. Every read request is forwarded to remote-read. The alerting & recording rules evaluation is performed by querying the data from remote-read. 

If disabled the data from promethues will be used. In both alerting & recording rules the evaluation happens for every `15s` or can be configured and the new series are written to promscale. The data retention policies configured should be higher than the time interval specified in alerting & recording rules (only if `read_recent` is disabled.)

Reviewer: Then let's mention users use `read_recent: true` as recommended over here and in the `note` section.

Adding recording rules:

```
groups:
  - name: my-rules
    rules:
    - record: <RECORD_NAME>
      expr: <PROMQL_EXPR>
      
```

Load the above defined recording rule file to your prometheus config for evaluation. i.e 

```
rule_files:
  - "<recording-rules-file>"
```

> More details on recording rules can be found [here](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/)

Reviewer: ```suggestion
More details on recording rules can be found [here](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/).
```

```
groups:
  - name: <alert-group-name>
    rules:
    - alert: <alert-name>
      expr: <promql_expression>
      for: <time-interval for how long this to happen to happen to fire an alert>
      labels:
        <key>:<value>
      annotations:
        summary: <text>
        description: <description on alert>
```

Load the above defined alert rule file to your prometheus configuration for evaluation.

```
rule_files:
  - "<alert-rules-file>"
```

Configure the alert manager in prometheus configuration.

```
alerting:
  alertmanagers:
  - static_configs:
    - targets:
      - localhost:9093
```

More details on alerting rules can be found [here](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/).

### Recording Rules

> The recording rules are for pre-calculating frequently used or computationally expensive queries. Recording rules are evaluated at the Prometheus side i.e., defined in recording rules in the Prometheus configuration. The evaluated rules form a new series which go by the name of the record field's value and ingested in both Prometheus and Promscale.

Reviewer: ```suggestion
The recording rules are for pre-calculating frequently used or computationally expensive queries. PromQL expressions defined in recording rules are evaluated at the Prometheus side. These evaluated rules form a new series that goes by the name of the record field's value and ingested in both Prometheus and Promscale.
```

Adding recording rules:

```
groups:
  - name: my-rules
    rules:
    - record: <RECORD_NAME>
      expr: <PROMQL_EXPR>
      
```

Load the above defined recording rule file to your prometheus configuration for evaluation.

```
rule_files:
  - "<recording-rules-file>"
```

More details on recording rules can be found [here](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/).

> **Note**: If `read_recent: true` is set in prometheus configuration the data is fetched from promscale for evaluation. If `read_recent` is disabled the data stored in prometheus will be used for evaluation this needs prometheus retention to be configured higher than the alerting & recording rules evaluation interval for the rules evaluation to happen before dropping the samples. The new alerting & recording rules series are written to promscale.  

Reviewer: Let's mention here that `read_recent: true` is the recommended way.

Reviewer: This statement can be modified to sound better. Also, the second sentence is too long.

> **Note**: The recommended way is to enable `read_recent` to `true` in the Prometheus configuration. This lets you fetch the data from Promscale for evaluation of PromQL queries. If `read_recent` is disabled, the data stored in Prometheus's local tsdb will be used for evaluating alerting/recording rules. Evaluation of rules on Prometheus's local tsdb will have dependency on Prometheus's retention time and not by Promscale.

Reviewer 2: ```suggestion
**Note**: We recommended setting `read_recent` to `true` in the [Prometheus remote_read configuration](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_read) when using recording rules. This tells Prometheus to fetch data from Promscale when evaluating PromQL queries (including recording rules). If `read_recent` is disabled, **only** the data stored in Prometheus's local tsdb will be used when evaluating alerting/recording rules and thus be dependent on the retention period of Prometheus (not Promscale).
```
